src/group.rs

```diff
@@ -163,3 +163,509 @@ impl Group for RistrettoPoint {
+        .unwrap();
+        // `L: 48`
+        const L: usize = 48;
+        // `Z: -10`
+        let z = BigInt::from(-10);
+
+        // extract points
+        let u0 = BigInt::from_bytes_be(Sign::Plus, &uniform_bytes[0..L]);
+        let u1 = BigInt::from_bytes_be(Sign::Plus, &uniform_bytes[L..L * 2]);
+
+        // map to curve
+        let (q0x, q0y) = map_to_curve_simple_swu(&u0, &a, &b, &p, &z);
+        let (q1x, q1y) = map_to_curve_simple_swu(&u1, &a, &b, &p, &z);
+
+        // convert to `p256` types
+        let p0 = AffinePoint::from_encoded_point(&EncodedPoint::from_affine_coordinates(
+            q0x.to_bytes_be().1[..].into(),
+            q0y.to_bytes_be().1[..].into(),
+            false,
+        ))
+        .ok_or(InternalPakeError::PointError)?
+        .to_curve();
+        let p1 = AffinePoint::from_encoded_point(&EncodedPoint::from_affine_coordinates(
+            q1x.to_bytes_be().1[..].into(),
+            q1y.to_bytes_be().1[..].into(),
+            false,
+        ))
+        .ok_or(InternalPakeError::PointError)?;
+
+        Ok(p0 + p1)
+    }
+
+    fn base_point() -> Self {
+        Self::generator()
+    }
+
+    fn mult_by_slice(&self, scalar: &GenericArray<u8, Self::ScalarLen>) -> Self {
+        self * &Self::Scalar::from_bytes_reduced(scalar)
+    }
+
+    fn is_identity(&self) -> bool {
+        self == &Self::identity()
+    }
+    fn ct_equal(&self, other: &Self) -> bool {
+        constant_time_eq::constant_time_eq(&self.to_arr(), &other.to_arr())
+    }
+}
+
+#[cfg(feature = "p256")]
+/// Corresponds to the map_to_curve_simple_swu() function defined in
+/// <https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#appendix-F.2>
+#[allow(clippy::many_single_char_names)]
+fn map_to_curve_simple_swu(
+    u: &num_bigint::BigInt,
+    a: &num_bigint::BigInt,
+    b: &num_bigint::BigInt,
+    p: &num_bigint::BigInt,
+    z: &num_bigint::BigInt,
+) -> (num_bigint::BigInt, num_bigint::BigInt) {
+    use num_bigint::BigInt;
+    use num_integer::Integer;
+    use num_traits::One;
+    use num_traits::ToPrimitive;
+    use std::ops::{Add, Div, Neg, Sub};
+
+    #[derive(Clone)]
+    struct Field<'a>(&'a BigInt);
+
+    impl<'a> Field<'a> {
+        fn new(p: &'a BigInt) -> Self {
+            Self(p)
+        }
+
+        fn element(&'a self, number: &BigInt) -> FieldElement<'a> {
+            FieldElement {
+                number: number.mod_floor(self.0),
+                f: self,
+            }
+        }
+
+        fn one(&'a self) -> FieldElement<'a> {
+            self.element(&BigInt::one())
+        }
+
+        /// See <https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-4>
+        fn inv0(&'a self, number: &FieldElement<'a>) -> FieldElement<'a> {
+            number.pow_internal(&(self.0 - 2))
+        }
+    }
+
+    /// Finite field arithmetic
+    #[derive(Clone)]
+    struct FieldElement<'a> {
+        number: BigInt,
+        f: &'a Field<'a>,
+    }
```

Comment:
I would have preferred not to implement something like this by hand, but I can't seem to find an appropriate library.

Reply:
This is an impressive endeavor! Perhaps it would be worth it to split out a bunch of this code into a separate file (like p256.rs)?

Reply:
I hope the way I split up the modules now is acceptable?